### ISSUE TITLE
Add a sky coordinate parser?

### ISSUE THREAD

**Issue by cdeil** [astropy/astropy#6379]
I want a sky coordinate parser, i.e. a function that takes a single string and makes a `SkyCoord` from it. This is partly possible with `SkyCoord` directly, e.g. this works:
```python
>>>  SkyCoord('42h 43d')
<SkyCoord (ICRS): (ra, dec) in deg
    ( 270.,  43.)>
```
But I couldn't find a way to make a Galactic frame SkyCoord, e.g. for to select the Galactic center I'd like to be able to write something like:
```python
>>> SkyCoord.parse('0 0 galactic')
```

The use cases I have for this are simple-to-use end-user interfaces like
https://github.com/hipspy/hips/issues/82 or config files and GUIs where one can simply select a sky position.

So the questions I have:

1. Does this exist in Astropy core and I just don't know about it?
2. Would something like `SkyCoord.parse` be welcome? (I envision a 10 line function that splits the string "lon lat frame" and then dispatches to existing angle parsing and is limited in what frames and frame attributes it supports)
3. Does someone know of an astronomy standard or spec or good existing example of such a parser? (I only found http://simbad.u-strasbg.fr/simbad/sim-fcoo and I think it's similar to the `SkyCoord` parser, doesn't support selection of a frame within the string).

If it's not considered appropriate for Astropy core, I would start by implementing a utility function for https://github.com/hipspy/hips/issues/82 and put it there in `hips.utils` (and also write a Javascript version for a website I'm working on).

**Comment by Cadair** [comment]
SunPy would also find this very useful. It would be good if each frame could register a pattern with SkyCoord to use. I.e. the galactic frame might register `'{lat} {lon} galatic'`. The solar conventions for representing coordinates as strings are complex so I would like this to be pretty flexible. Maybe allowing each frame to register a matching regex and then a method to handle the parseing?

**Comment by adrn** [comment]
This is related to (or maybe a duplicate of) the ancient #1115 -- see later in the thread for discussion.

cc @bsipocz

**Comment by Cadair** [comment]
I think this is probably a superset of #1115 from a quick read. My main interest is to have a pluggable way of defining string parsers in SkyCoord, which looks a little different from what's in that issue.

**Comment by taldcroft** [comment]
I'm hearing two things:
- @cdeil wants the specific ability to include the frame name at the end of a single string
- @Cadair wants to support arbitrary patterns defined programmatically

The first is likely quite easy by inserting a few lines of code in the existing [SkyCoord._parse_inputs](https://github.com/astropy/astropy/blob/8467c8728e0d6281aad60547df2af305faff29f2/astropy/coordinates/sky_coordinate.py#L311) method, probably near the top.  This could basically look like:
```
if len(args) == 1:
    arg_frame, new_args0 = extract_frame_name(args[0])
    if arg_frame:
        frame = arg_frame
        args[0] = new_args0  # extract_frame_name has removed the frame string from args[0]
```
Implementation is easy but the harder question is whether this is sufficiently general to be of broad use.  @cdeil touched on this by asking if there is any prior art, and the point is to be careful about inventing ad-hoc schemes for representing coordinates.

The request from @Cadair is a much bigger question that touches on the general idea of having format specifications in coordinates.  The discussion in #1115 is just the tip of the iceberg on that topic.  There was a [long and rather lively discussion](https://groups.google.com/d/msg/astropy-dev/xw-xOv_THWQ/w3p9CChvo6IJ) in the past on the topic on astropy-dev which in the end went nowhere.   This was focused on output formatting, but in my mind anyway the same could be applied to the input format, much like how `Time` works.

In theory one could punt on the whole concept of a format language and let Frame subclasses define their own parser method that gets called at the right spot in `SkyCoord._parse_inputs`.

**Comment by cdeil** [comment]
> the harder question is whether this is sufficiently general to be of broad use. @cdeil touched on this by asking if there is any prior art, and the point is to be careful about inventing ad-hoc schemes for representing coordinates.

Yes, it would be great if there were a scheme that many astronomy tools support and we should probably try hard not to invent our own mini language that no other tool supports (unless there is no good prior art on this, which I doubt).

So if someone knows of a standard or spec, or at least a description or open-source code what other astronomy tools support, please comment or link to it from here.

**Comment by Cadair** [comment]
> In theory one could punt on the whole concept of a format language and let Frame subclasses define their own parser method that gets called at the right spot in `SkyCoord._parse_inputs`.

This please! 

The only question is if you require users to explicitly set the frame, or if we do frame detection based on the string input. It felt to me like @cdeil was asking for frame detection based on the string input. I was suggesting that maybe a more generalised version of that would be useful for us.

(For a little context one of the things we want to be able to support is described in [this](https://link.springer.com/article/10.1007/s11207-010-9553-0) paper, which is strings that look like this `'SOLyyyy-mm-ddThh:mm:ssLdddCdddRnnn.f'`)
